```
[2019-10-01T17:27:18.679Z] Building test suite binary /go/src/github.com/docker/docker/integration/service/test.main

[2019-10-01T17:27:18.937Z] # github.com/docker/docker/daemon

[2019-10-01T17:27:18.937Z] daemon/start.go:185:51: undefined: c

[2019-10-01T17:27:18.937Z] daemon/start.go:190:86: cannot use imageName (type "github.com/docker/docker/vendor/github.com/docker/distribution/reference".Named) as type string in argument to withImageName

[2019-10-01T17:27:18.937Z] daemon/start.go:199:88: cannot use imageName (type "github.com/docker/docker/vendor/github.com/docker/distribution/reference".Named) as type string in argument to withImageName

[2019-10-01T17:27:18.937Z] daemon/start.go:278:9: cannot use func literal (type func(context.Context, *"github.com/docker/docker/vendor/github.com/containerd/containerd".Client, "github.com/docker/docker/vendor/github.com/containerd/containerd/containers".Container) error) as type "github.com/docker/docker/vendor/github.com/containerd/containerd".NewContainerOpts in return argument

[2019-10-01T17:27:22.229Z] # github.com/docker/docker/daemon

[2019-10-01T17:27:22.229Z] daemon/start.go:185:51: undefined: c

[2019-10-01T17:27:22.229Z] daemon/start.go:190:86: cannot use imageName (type "github.com/docker/docker/vendor/github.com/docker/distribution/reference".Named) as type string in argument to withImageName

[2019-10-01T17:27:22.229Z] daemon/start.go:199:88: cannot use imageName (type "github.com/docker/docker/vendor/github.com/docker/distribution/reference".Named) as type string in argument to withImageName

[2019-10-01T17:27:22.229Z] daemon/start.go:278:9: cannot use func literal (type func(context.Context, *"github.com/docker/docker/vendor/github.com/containerd/containerd".Client, "github.com/docker/docker/vendor/github.com/containerd/containerd/containers".Container) error) as type "github.com/docker/docker/vendor/github.com/containerd/containerd".NewContainerOpts in return argument
```

--- a/daemon/start.go
+++ b/daemon/start.go
@@ -5,6 +5,9 @@ import (
 	"runtime"
 	"time"
 
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/containers"
+	"github.com/docker/distribution/reference"
 	"github.com/docker/docker/api/types"
 	containertypes "github.com/docker/docker/api/types/container"
 	"github.com/docker/docker/container"
@@ -179,7 +182,12 @@ func (daemon *Daemon) containerStart(container *container.Container, checkpoint 
 
 	ctx := context.TODO()
 
-	err = daemon.containerd.Create(ctx, container.ID, spec, createOptions)
+	imageRef, err := reference.ParseNormalizedNamed(container.Config.Image)
+	if err != nil {
+		return err
+	}
+
+	err = daemon.containerd.Create(ctx, container.ID, spec, createOptions, withImageName(imageRef.String()))
 	if err != nil {
 		if errdefs.IsConflict(err) {
 			logrus.WithError(err).WithField("container", container.ID).Error("Container not cleaned up from containerd from previous run")
@@ -188,7 +196,7 @@ func (daemon *Daemon) containerStart(container *container.Container, checkpoint 
 			if err := daemon.containerd.Delete(ctx, container.ID); err != nil && !errdefs.IsNotFound(err) {
 				logrus.WithError(err).WithField("container", container.ID).Error("Error cleaning up stale containerd container object")
 			}
-			err = daemon.containerd.Create(ctx, container.ID, spec, createOptions)
+			err = daemon.containerd.Create(ctx, container.ID, spec, createOptions, withImageName(imageRef.String()))
 		}
 		if err != nil {
 			return translateContainerdStartErr(container.Path, container.SetExitCode, err)
@@ -263,5 +271,12 @@ func (daemon *Daemon) Cleanup(container *container.Container) {
 
 	if err := daemon.containerd.Delete(context.Background(), container.ID); err != nil {
 		logrus.Errorf("%s cleanup: failed to delete container from containerd: %v", container.ID, err)
+	}
+}
+
+func withImageName(n string) containerd.NewContainerOpts {
+	return func(ctx context.Context, _ *containerd.Client, c *containers.Container) error {
+		c.Image = n
+		return nil
 	}
 }

--- a/daemon/util_test.go
+++ b/daemon/util_test.go
@@ -28,7 +28,7 @@ func (c *MockContainerdClient) Version(ctx context.Context) (containerd.Version,
 func (c *MockContainerdClient) Restore(ctx context.Context, containerID string, attachStdio libcontainerdtypes.StdioCallback) (alive bool, pid int, p libcontainerdtypes.Process, err error) {
 	return false, 0, &mockProcess{}, nil
 }
-func (c *MockContainerdClient) Create(ctx context.Context, containerID string, spec *specs.Spec, runtimeOptions interface{}) error {
+func (c *MockContainerdClient) Create(ctx context.Context, containerID string, spec *specs.Spec, runtimeOptions interface{}, opts ...containerd.NewContainerOpts) error {
 	return nil
 }
 func (c *MockContainerdClient) Start(ctx context.Context, containerID, checkpointDir string, withStdin bool, attachStdio libcontainerdtypes.StdioCallback) (pid int, err error) {

--- a/libcontainerd/local/local_windows.go
+++ b/libcontainerd/local/local_windows.go
@@ -152,7 +152,7 @@ func (c *client) Version(ctx context.Context) (containerd.Version, error) {
 //		"ImagePath": "C:\\\\control\\\\windowsfilter\\\\65bf96e5760a09edf1790cb229e2dfb2dbd0fcdc0bf7451bae099106bfbfea0c\\\\UtilityVM"
 //	},
 //}
-func (c *client) Create(_ context.Context, id string, spec *specs.Spec, runtimeOptions interface{}) error {
+func (c *client) Create(_ context.Context, id string, spec *specs.Spec, runtimeOptions interface{}, opts ...containerd.NewContainerOpts) error {
 	if ctr := c.getContainer(id); ctr != nil {
 		return errors.WithStack(errdefs.Conflict(errors.New("id already in use")))
 	}

--- a/libcontainerd/remote/client.go
+++ b/libcontainerd/remote/client.go
@@ -124,15 +124,18 @@ func (c *client) Restore(ctx context.Context, id string, attachStdio libcontaine
 	}, nil
 }
 
-func (c *client) Create(ctx context.Context, id string, ociSpec *specs.Spec, runtimeOptions interface{}) error {
+func (c *client) Create(ctx context.Context, id string, ociSpec *specs.Spec, runtimeOptions interface{}, opts ...containerd.NewContainerOpts) error {
 	bdir := c.bundleDir(id)
 	c.logger.WithField("bundle", bdir).WithField("root", ociSpec.Root.Path).Debug("bundle dir created")
 
-	_, err := c.client.NewContainer(ctx, id,
+	newOpts := []containerd.NewContainerOpts{
 		containerd.WithSpec(ociSpec),
 		containerd.WithRuntime(runtimeName, runtimeOptions),
 		WithBundle(bdir, ociSpec),
-	)
+	}
+	opts = append(opts, newOpts...)
+
+	_, err := c.client.NewContainer(ctx, id, opts...)
 	if err != nil {
 		if containerderrors.IsAlreadyExists(err) {
 			return errors.WithStack(errdefs.Conflict(errors.New("id already in use")))

--- a/libcontainerd/types/types.go
+++ b/libcontainerd/types/types.go
@@ -52,7 +52,7 @@ type Client interface {
 
 	Restore(ctx context.Context, containerID string, attachStdio StdioCallback) (alive bool, pid int, p Process, err error)
 
-	Create(ctx context.Context, containerID string, spec *specs.Spec, runtimeOptions interface{}) error
+	Create(ctx context.Context, containerID string, spec *specs.Spec, runtimeOptions interface{}, opts ...containerd.NewContainerOpts) error
 	Start(ctx context.Context, containerID, checkpointDir string, withStdin bool, attachStdio StdioCallback) (pid int, err error)
 	SignalProcess(ctx context.Context, containerID, processID string, signal int) error
 	Exec(ctx context.Context, containerID, processID string, spec *specs.Process, withStdin bool, attachStdio StdioCallback) (int, error)

